ecs-cli/modules/utils/compose/ecs_params_reader_test.go

```diff
@@ -214,6 +214,68 @@ run_params:
 	}
 }
 
+func TestReadECSParams_WithTaskPlacement(t *testing.T) {
+	ecsParamsString := `version: 1
+run_params:
+  task_placement:
+    strategy:
+      - field: memory
+        type: binpack
+      - field: attribute:ecs.availability-zone
+        type: spread
+    constraints:
+      - expression: attribute:ecs.instance-type =~ t2.*
```

Comment:
Can we add a few more examples of the query language to this test to make sure they're all read in appropriately?

Reply:
`expression` is just a string -- there isn't any special parsing logic for what the string looks like. Since our philosophy in the CLI has been to defer validation to the service, there's not too much value in lots of examples of valid query strings since that all happens in ECS. If it's for purposes of documentation, I have a few more in the `README.md` that also talks about task group constraints.

Reply:
I'm less concerned with validating that the input makes sense (as you point out, the service will return an error), and more concerned with making sure we're correctly reading/unmarshalling special chars that a customer may want to use in an expression.

Reply:
Verifying "!" special char
```
~/ecs/ecs-cli/testing/projects/test_fields$ cat ./ecs_params/ecs-params-task-placement.yml 
version: 1
task_definition:
run_params:
  task_placement:
    strategy:
      - field: memory
        type: binpack
    constraints:
      - expression: attribute:color != green
        type: memberOf

~/ecs/ecs-cli/testing/projects/test_fields$ ecs-cli compose  --ecs-params ecs_params/ecs-params-task-placement.yml --cluster test-attributes --region us-west-2 service up
INFO[0000] Using ECS task definition                     TaskDefinition="test_fields:56"
INFO[0000] Created an ECS service                        service=test_fields taskDefinition="test_fields:56"
INFO[0000] Updated ECS service successfully              desiredCount=1 serviceName=test_fields
INFO[0015] (service test_fields) has started 1 tasks: (task 73641fe0-c692-48f5-b927-e81f3a82c124).  timestamp="2018-08-21 22:20:17 +0000 UTC"
INFO[0030] Service status                                desiredCount=1 runningCount=1 serviceName=test_fields
INFO[0030] ECS Service has reached a stable state        desiredCount=1 runningCount=1 serviceName=test_fields

~/ecs/ecs-cli/testing/projects/test_fields$ aws ecs describe-services --services test_fields --cluster test-attributes
{
    "services": [
        {
            "status": "ACTIVE", 
            "taskDefinition": "arn:aws:ecs:us-west-2:xxxxxxxxxx:task-definition/test_fields:56", 
            "pendingCount": 0, 
            "launchType": "EC2", 
            "loadBalancers": [], 
            "placementConstraints": [
                {
                    "expression": "attribute:color != green", 
                    "type": "memberOf"
                }
            ], 
            "createdAt": 1534890014.687, 
            "desiredCount": 1, 
            "serviceName": "test_fields", 
          .... // bunch of other stuff
            "placementStrategy": [
                {
                    "field": "MEMORY", 
                    "type": "binpack"
                }
            ]
        }
    ], 
    "failures": []
}
```
Verifying `<` special char:
```
version: 1
task_definition:
run_params:
  task_placement:
    strategy:
      - field: memory
        type: binpack
    constraints:
      - expression: attribute:color <= green
        type: memberOf


// compose service up

~/ecs/ecs-cli/testing/projects/test_fields$ aws ecs describe-services --services test_fields --cluster test-attributes
{
    "services": [
        {
            "status": "ACTIVE", 
            "taskDefinition": "arn:aws:ecs:us-west-2:794715269151:task-definition/test_fields:56", 
            "pendingCount": 0, 
            "launchType": "EC2", 
            "loadBalancers": [], 
            "placementConstraints": [
                {
                    "expression": "attribute:color <= green", 
                    "type": "memberOf"
                }
            ], 
```

Reply:
Thanks!